### PR TITLE
Add documentation on the URL structure for deeplinking📘 #2553

### PR DESCRIPTION
**Pull request by robertbasti**
[//]: # (Request Template Witsml Explorer)

[//]: # (Thank you for contributing to WITSML Explorer! Before submitting this PR, please fill in the following template.)

## Fixes

[//]: #2553

This pull request fixes issue #2553 

## Description

A documentation how to create deep links from external applications

## Type of change

[//]: # (Mark any of the types of change that apply.)

* [ ] Bugfix
* [ ] New feature (non-breaking change which adds functionality)
* [ ] Enhancement of existing functionality
* [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
* [x] This change requires a documentation update

## Impacted Areas in Application

[//]: # (List general components of the application that this PR will affect)

* [ ] Frontend
* [ ] API
* [ ] WITSML
* [ ] Desktop
* [x] Other (documentation)

## Checklist:

[//]: # (Please tick all the boxes or remove the ones that aren't needed and explain why)

*Communication*
* [ ] I have made corresponding changes to the documentation
* [ ] PR affects application security

*Code quality*
* [ ] I have self-reviewed my code
* [ ] No new warnings are generated

*Test coverage*
* [ ] New code is covered by passing tests

## Further comments

[//]: # (If this is a relatively large or complex change, kick off the discussion by explaining why you chose the solution you did and what alternatives you considered, etc...)

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -49,6 +49,9 @@ Please visit [API client access](/Docs/APICLIENT.md).
 ## Enable HTTP WITSML servers
 Please see [enable HTTP](/Docs/enable_http.md).
 
+## URL structure for deeplinking
+Please see [enable HTTP](/Docs/deep_linking.md).
```
Comment (eliasbruvik) [inline README.md]: enable HTTP -> deep linking

File: Docs/deep_linking.md [hunk 2]
```diff
@@ -0,0 +1,78 @@
...
+/servers/:serverUrl/wells/:wellUid/wellbores/:wellboreUid/objectgroups/:objectGroup/objects
+
+### Object View
+/servers/:serverUrl/wells/:wellUid/wellbores/:wellboreUid/objectgroups/:objectGroup/objects/:objectUid
+
+### Log Types View
+/servers/:serverUrl/wells/:wellUid/wellbores/:wellboreUid/objectgroups/:objectGroup/logtypes
+
+### Log Objects View
+/servers/:serverUrl/wells/:wellUid/wellbores/:wellboreUid/objectgroups/:objectGroup/logtypes/:logType/objects/:objectUid
```
Comment (eliasbruvik) [inline Docs/deep_linking.md]: remove `/:objectUid` from this one

File: Docs/deep_linking.md [hunk 3]
```diff
@@ -0,0 +1,78 @@
...
+
+### Log Object View
+/servers/:serverUrl/wells/:wellUid/wellbores/:wellboreUid/objectgroups/:objectGroup/logtypes/:logType/objects/:objectUid/curvevalues
```
Comment (eliasbruvik) [inline Docs/deep_linking.md]: Remove `/curvevalues` from this one

File: Docs/deep_linking.md [hunk 4]
```diff
@@ -0,0 +1,78 @@
...
+This paramater can have values:
+  - time
+  - depth
+
+### filterType
+
+This paramater can have values:
+  - Log
+  - Group
+  - ServiceCompany
```
Comment (eliasbruvik) [inline Docs/deep_linking.md]: Seems incorrect. The values can be Wellbore, Log, Rig or Service%20Company.

File: Docs/deep_linking.md [hunk 5]
```diff
@@ -0,0 +1,78 @@
+# URL structure for deeplinking
+
+ This describes the structure of the URL so that it is easy for developers to create deep links from other applications directly to a well/wellbore/object/curve data etc.
+
```
Comment (eliasbruvik) [inline Docs/deep_linking.md:4]: Add a section about that all server and uid parameters must be URL encoded.

File: Docs/deep_linking.md
Comment (eliasbruvik) [inline Docs/deep_linking.md:1]: We also need to document the available search params. 
* Most pages can receive a filter to filter the rows of the table in that page (for example `?filter=%7B"name"%3A"test"%7D`).
* Need to explain the format of search params for the curve values view (for example `?startIndex=1&endIndex=2&mnemonics=%5B"Curve1"%2C"Curve2"%5D'`)
* Need to explain the format of search params for the multi logs (for example `?logs=%5B"bigDepthLog"%2C"72b5ec0c-ddd7-4ba6-a8bd-114c0d8d9f2a"%5D`)
  * Same goes for multi curve values view (for example `?startIndex=0&endIndex=5&mnemonics=%7B"bigDepthLog"%3A%5B"A"%2C"B"%2C"C"%5D%7D`)

**Review by eliasbruvik — CHANGES_REQUESTED** [review]

File: README.md (hunk 1 above)
Comment (robertbasti) [reply README.md]: fixed

File: Docs/deep_linking.md (hunk 2 above)
Comment (robertbasti) [reply Docs/deep_linking.md]: done

File: Docs/deep_linking.md (hunk 3 above)
Comment (robertbasti) [reply Docs/deep_linking.md]: done

File: Docs/deep_linking.md (hunk 4 above)
Comment (robertbasti) [reply Docs/deep_linking.md]: done

File: Docs/deep_linking.md (hunk 5 above)
Comment (robertbasti) [reply Docs/deep_linking.md:4]: done

File: Docs/deep_linking.md
Comment (robertbasti) [reply Docs/deep_linking.md:1]: solved

File: Docs/deep_linking.md [hunk 6]
```diff
@@ -0,0 +1,169 @@
...
+### Filtering Rows Of Views
+
+Most pages can receive a filter to filter the rows of the table in that page. Those filters are added at the end of the url.
+For example showing only rows where parameter "name" has value "test":
+
+?filter={"name":"test"}
+
+Url encoded version which needs to be used:
+
+%3Ffilter%3D%7B%22name%22%3A%22test%22%7D
```
Comment (eliasbruvik) [inline Docs/deep_linking.md]: `?filter=` should not be encoded.

File: Docs/deep_linking.md [hunk 7]
```diff
@@ -0,0 +1,169 @@
...
+
+
+### Log Curve Values View
+/servers/:serverUrl/wells/:wellUid/wellbores/:wellboreUid/objectgroups/:objectGroup/logtypes/:logType/objects/:objectUid/curvevalues
```
Comment (eliasbruvik) [inline Docs/deep_linking.md]: Maybe add something to the end here so it's easier to understand that the search params are expected (but not required)? Like `:objectUid/curvevalues?startIndex=<start index>&endIndex=<end index>&mnemonics=<mnemonics>`

File: Docs/deep_linking.md [hunk 8]
```diff
@@ -0,0 +1,169 @@
...
+
+#### Additional params for searching
```
Comment (eliasbruvik) [inline Docs/deep_linking.md]: The params in this case are not for searching, but for limiting the data to the given range and mnemonics.

File: Docs/deep_linking.md [hunk 9]
```diff
@@ -0,0 +1,169 @@
...
+
+If there is a need to search in result, following parameters need to be filled up:
```
Comment (eliasbruvik) [inline Docs/deep_linking.md]: See comment above. Saying it's a search is very confusing.

File: Docs/deep_linking.md [hunk 10]
```diff
@@ -0,0 +1,169 @@
...
+
+StartIndex and endIndex could be numbers for depth based logs or datetime in ISO format  - "YYYY-MM-DDTHH:mm:ss.sssZ" for time based logs
+
+Samples:
+
+##### Time sample:
+?startIndex=2024-09-06T00:00:00Z&endIndex=2024-09-06T00:01:00Z&mnemonics=["mnem1","mnem2","mnem3","mnem4"]
+
+url encoded version which should be used:
+%3FstartIndex%3D2024-09-06T00%3A00%3A00Z%26endIndex%3D2024-09-06T00%3A01%3A00Z%26mnemonics%3D%5B%22mnem1%22%2C%22mnem2%22%2C%22mnem3%22%2C%22mnem4%22%5D
```
Comment (eliasbruvik) [inline Docs/deep_linking.md]: Each search param value must be encoded by it's own. `?startIndex=`, `&endIndex=` and `&mnemonics=` should not be encoded.

File: Docs/deep_linking.md
Comment (eliasbruvik) [inline Docs/deep_linking.md:1]: Add a comment somewhere that the full url should not exceed the maximum url length accepted by most browsers. I think we have a cutoff of 2000 characters in WEx. This mostly applies to if you select a lot of mnemonics in the curve values view (or multi curve values view), so maybe it should be added there.

File: Docs/deep_linking.md [hunk 11]
```diff
@@ -0,0 +1,169 @@
...
+
+In the sample above we combine mnemonics "mmem1", "mmem2", "mmem3", "mmen4" for the selected time interval
+
+##### Depth sample:
+
+?startIndex=1&endIndex=4&mnemonics=["mnem1","mnem2","mnem3","mnem4"]
+
+url encoded version which should be used:
+%3FstartIndex%3D1%26endIndex%3D4%26mnemonics%3D%5B%22mnem1%22%2C%22mnem2%22%2C%22mnem3%22%2C%22mnem4%22%5D
```
Comment (eliasbruvik) [inline Docs/deep_linking.md]: Same as above

File: Docs/deep_linking.md [hunk 12]
```diff
@@ -0,0 +1,169 @@
...
+
+### Multi Logs Curve Info List View
+/servers/:serverUrl/wells/:wellUid/wellbores/:wellboreUid/objectgroups/:objectGroup/logtypes/:logType/multilogs"
```
Comment (eliasbruvik) [inline Docs/deep_linking.md]: Add the search params expected here also (like `?logs=<log uids>`)

File: Docs/deep_linking.md [hunk 13]
```diff
@@ -0,0 +1,169 @@
...
+
+
+#### Additional params for searching
```
Comment (eliasbruvik) [inline Docs/deep_linking.md]: Again, not for searching

File: Docs/deep_linking.md [hunk 14]
```diff
@@ -0,0 +1,169 @@
...
+
+If there is a need to search in result, following parameter needs to be filled up:
```
Comment (eliasbruvik) [inline Docs/deep_linking.md]: Not for searching

File: Docs/deep_linking.md [hunk 15]
```diff
@@ -0,0 +1,169 @@
...
+ 
+and url-encoded version, which should be used:
+
+?logs=%5B%22323f827b-c849-4b14-a1c3-ba207848f29c%22%2C%22701f5f00-bdb5-4937-8350-1975240ce432%22%5D
+
+In the sample above we combine 2 logs with uid "323f827b-c849-4b14-a1c3-ba207848f29c" and with uid "701f5f00-bdb5-4937-8350-1975240ce432".
+
+
+### Multi Logs Curve Values
+/servers/:serverUrl/wells/:wellUid/wellbores/:wellboreUid/objectgroups/:objectGroup/logtypes/:logType/multilogs/curvevalues"
```
Comment (eliasbruvik) [inline Docs/deep_linking.md]: Add the search params expected here also (like `?startIndex=<start index>&endIndex=<end index>&mnemonics=<mnemonics for each log>`)

File: Docs/deep_linking.md [hunk 16]
```diff
@@ -0,0 +1,169 @@
...
+
+#### Additional params for searching
+
+If there is a need to search in result, following parameters need to be filled up:
```
Comment (eliasbruvik) [inline Docs/deep_linking.md]: Not search

File: Docs/deep_linking.md [hunk 17]
```diff
@@ -0,0 +1,169 @@
...
+
+StartIndex and endIndex could be numbers for depth based logs or datetime in ISO format  - "YYYY-MM-DDTHH:mm:ss.sssZ" for time based logs
+
+Samples:
+
+##### Time sample:
+?startIndex=2024-09-06T00:00:00.000Z&endIndex=2024-09-06T00:01:00.000Z&mnemonics={"323f827b-c849-4b14-a1c3-ba207848f29c":["mnem1"],"701f5f00-bdb5-4937-8350-1975240ce432":["mnem2"]}
+
+url encoded version which should be used:
+%3FstartIndex%3D2024-09-06T00%3A00%3A00.000Z%26endIndex%3D2024-09-06T00%3A01%3A00.000Z%26mnemonics%3D%7B%22323f827b-c849-4b14-a1c3-ba207848f29c%22%3A%5B%22mnem1%22%5D%2C%22701f5f00-bdb5-4937-8350-1975240ce432%22%3A%5B%22mnem2%22%5D%7D
```
Comment (eliasbruvik) [inline Docs/deep_linking.md]: Remove url-encoding of parts of this (like mentioned previously)

File: Docs/deep_linking.md [hunk 18]
```diff
@@ -0,0 +1,169 @@
...
+
+In the sample above we combine mnemonics "mmem1" from log with uid "323f827b-c849-4b14-a1c3-ba207848f29c" and "mmem2" from log with uid "701f5f00-bdb5-4937-8350-1975240ce432".
+
+##### Depth sample:
+
+?1&endIndex=4&mnemonics={"55823e08-6232-43a8-8e77713bc469-6674":["mnem3"],"bc498d56-84f6-4c63-b1af-d845be3e6961":["mnem4"]}
+
+url encoded version which should be used:
+%3F1%26endIndex%3D4%26mnemonics%3D%7B%2255823e08-6232-43a8-8e77713bc469-6674%22%3A%5B%22mnem3%22%5D%2C%22bc498d56-84f6-4c63-b1af-d845be3e6961%22%3A%5B%22mnem4%22%5D%7D
```
Comment (eliasbruvik) [inline Docs/deep_linking.md]: Remove url-encoding of parts of this (like mentioned previously)

File: Docs/deep_linking.md
Comment (eliasbruvik) [inline Docs/deep_linking.md:1]: It's a bit hard to read all the examples (both before and after encoding). Could you try to format them as code blocks or find another way to make it look better? Currently they just look like normal text.

**Review by eliasbruvik — CHANGES_REQUESTED** [review]

File: Docs/deep_linking.md (hunk 6 above)
Comment (robertbasti) [reply Docs/deep_linking.md]: fixed

File: Docs/deep_linking.md (hunk 7 above)
Comment (robertbasti) [reply Docs/deep_linking.md]: fixed

File: Docs/deep_linking.md (hunk 8 above)
Comment (robertbasti) [reply Docs/deep_linking.md]: fixed

File: Docs/deep_linking.md (hunk 9 above)
Comment (robertbasti) [reply Docs/deep_linking.md]: fixed

File: Docs/deep_linking.md (hunk 10 above)
Comment (robertbasti) [reply Docs/deep_linking.md]: fixed

File: Docs/deep_linking.md (hunk 11 above)
Comment (robertbasti) [reply Docs/deep_linking.md]: fixed

File: Docs/deep_linking.md (hunk 12 above)
Comment (robertbasti) [reply Docs/deep_linking.md]: done

File: Docs/deep_linking.md (hunk 13 above)
Comment (robertbasti) [reply Docs/deep_linking.md]: done

File: Docs/deep_linking.md (hunk 14 above)
Comment (robertbasti) [reply Docs/deep_linking.md]: done

File: Docs/deep_linking.md (hunk 15 above)
Comment (robertbasti) [reply Docs/deep_linking.md]: done

File: Docs/deep_linking.md (hunk 16 above)
Comment (robertbasti) [reply Docs/deep_linking.md]: done

File: Docs/deep_linking.md (hunk 17 above)
Comment (robertbasti) [reply Docs/deep_linking.md]: done

File: Docs/deep_linking.md (hunk 18 above)
Comment (robertbasti) [reply Docs/deep_linking.md]: done

File: Docs/deep_linking.md
Comment (robertbasti) [reply Docs/deep_linking.md:1]: done

Comment (robertbasti) [reply Docs/deep_linking.md:1]: solved

File: Docs/deep_linking.md [hunk 19]
```diff
@@ -0,0 +1,186 @@
...
+```?startIndex=1&endIndex=4&mnemonics=%7B%2255823e08-6232-43a8-8e77713bc469-6674%22%3A%5B%22mnem3%22%5D%2C%22bc498d56-84f6-4c63-b1af-d845be3e6961%22%3A%5B%22mnem4%22%5D%7D%0A```
+
+### Jobs View
+/servers/:serverUrl/jobs
+
+### Query View
+/servers/:serverUrl/query
+
+### Search View
+/servers/:serverUrl/search/:filterType
```
Comment (eliasbruvik) [inline Docs/deep_linking.md]: Forgot formatting on these last 3

File: Docs/deep_linking.md [hunk 20]
```diff
@@ -0,0 +1,186 @@
...
+```?startIndex=1&endIndex=5&mnemonics=%5B%22mnem1%22%2C%22mnem2%22%2C%22mnem3%22%2C%22mnem4%22%5D```
+
+### Multi Logs Curve Info List View
+
+```/servers/:serverUrl/wells/:wellUid/wellbores/:wellboreUid/objectgroups/:objectGroup/logtypes/:logType/multilogs?logs=<log uids>```
+
+
+#### Additional params for limiting the data in result
+
+If there is a need to limit the in result, following parameter needs to be filled up:
```
Comment (eliasbruvik) [inline Docs/deep_linking.md]: I think this is required for multi-logs. And in the multi-log case it's maybe not as much to limit the results (as not selecting any value for logs does not implicitly use all logs like it does for the normal log curve values view).

**Review by eliasbruvik — CHANGES_REQUESTED** [review]

File: Docs/deep_linking.md (hunk 19 above)
Comment (robertbasti) [reply Docs/deep_linking.md]: fixed

File: Docs/deep_linking.md (hunk 20 above)
Comment (robertbasti) [reply Docs/deep_linking.md]: done

**Review by eliasbruvik — APPROVED** [review]

**Pull request merged**